### PR TITLE
#17 Add stress tests

### PR DESCRIPTION
**Pull request by Usipov**
Added some tests checking that `DifferenceKit` computes and applies diff to `UICollectionView` as expected: correctly and avoiding internal `UICollectionView` crashes.

To perform these tests I added a utility class catching Objective-C exceptions and passing them into Swift code.

Note, however, that these tests depend on `UIKit`, so CI integration should run tests not only on a Mac target, but also on a Simulator.

If your run the tests, you will see something like
```
testing iteration 93
  sections: from 9 to: 12.
  sections: deletes 4, inserts: 7, moves 4, updates: 3.
  cells:    from 2 to: 26.
  cells:    deletes 2, inserts: 26, moves 19, updates: 2.

testing iteration 94
  sections: from 12 to: 9.
  sections: deletes 7, inserts: 4, moves 5, updates: 4.
  cells:    from 26 to: 47.
  cells:    deletes 26, inserts: 47, moves 23, updates: 26.

testing iteration 95
  sections: from 9 to: 9.
  sections: deletes 3, inserts: 3, moves 2, updates: 3.
  cells:    from 47 to: 70.
  cells:    deletes 38, inserts: 61, moves 19, updates: 55.
```

### CODE REVIEW

File: Tests/StressTests/Support/CollectionViewUpdater.swift [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+        collection: Collection,
+        diff: StagedChangeset<Collection>,
+        completion: @escaping (CollectionViewUpdaterResult) -> ())
+    {
+        guard !diff.isEmpty else {
+            // If we apply an empty diff to a collection view, then for some reason
+            // `collectionView.indexPathsForVisibleItems` and `collectionView.visibleCells` return empty arrays,
+            // even if `collectionView.numberOfItems(inSection:)` returns a non empty number.
+            // So we cannot perform normal assertions as in `CollectionViewUpdaterResult.success`
+            return completion(.noUpdateRequired)
```
Comment (Usipov) [inline Tests/StressTests/Support/CollectionViewUpdater.swift]: This is a strange case. I have no idea, why this happens.
Same situation on `ListDiff` stress tests did not take place.
@ra1028, Do you have any thoughts about it?

**Comment by ra1028** [comment]
Thanks @Usipov !
The stress tests are works very well. It's awesome.
However, this test codes seems too large and no maintainability.
Could you make it minimal?
Or, Could you write the test focused on the certain cause of the crash?

File: Tests/StressTests/Support/CollectionViewUpdater.swift (hunk 1 above)
Comment (ra1028) [reply Tests/StressTests/Support/CollectionViewUpdater.swift]: Hmm, I'm not sure.
Nothing should happen in the empty StagedChangeset.

**Comment by Usipov** [comment]
Why do you think the code seems to be not maintainable?
I tried to follow SRP, and named things very carefully.

I cannot add a test for a particular crash, because I did not find any crashes in your library. That's the whole point... To make sure, the library is very relable.

You can revert your #18 and you will find test failures... Noone knew about that crash before my tests

**Comment by ra1028** [comment]
@Usipov
IMHO, I think that the amount of your code is too large(it's more than the diffing algorithm).
I want to keep the code base simple.
Nevertheless, I'd love to want to merge, because I believe that your test code is beneficial to DiffereceKit.

It's difficult to point out all in the review, so may I merge it into my working branch and refactor it?

**Comment by Usipov** [comment]
@ra1028, as you wish

**Comment by ra1028** [comment]
@Usipov Thanks 🙏

**Pull request merged**